src/main/java/com/commercetools/sync/customers/helpers/CustomerBatchValidator.java

```diff
@@ -0,0 +1,223 @@
+        for (int i = 0; i < list.size(); i++) {
+            if (predicate.test(list.get(i))) {
+                indexes.add(i);
+            }
+        }
+        return indexes;
+    }
+
+    private boolean hasValidBillingAndShippingAddresses(@Nonnull final CustomerDraft customerDraft) {
+        /* An example error response from the API, when the indexes are not valid:
+           {
+                "statusCode": 400,
+                "message": "Customer does not contain an address at index '1'.",
+                "errors": [{
+                    "code": "InvalidOperation",
+                    "message": "Customer does not contain an address at index '-1'."
+                },
+                {
+                    "code": "InvalidOperation",
+                    "message": "Customer does not contain an address at index '1'."
+                }]
+            }
+         */
+
+        final List<Address> addressList = customerDraft.getAddresses();
+        final Predicate<Integer> isInvalidIndex = index -> index == null || addressList == null
+            || addressList.isEmpty() || index < 0 || index > addressList.size() - 1;
+
+        if (customerDraft.getBillingAddresses() != null && !customerDraft.getBillingAddresses().isEmpty()) {
+            final List<Integer> invalidIndexes = getIndexValues(customerDraft.getBillingAddresses(), isInvalidIndex);
+            if (!invalidIndexes.isEmpty()) {
+                handleError(format(BILLING_ADDRESSES_ARE_NOT_VALID, customerDraft.getKey(), invalidIndexes));
+                return false;
+            }
+        }
+
+        if (customerDraft.getShippingAddresses() != null && !customerDraft.getShippingAddresses().isEmpty()) {
+            final List<Integer> invalidIndexes = getIndexValues(customerDraft.getShippingAddresses(), isInvalidIndex);
+            if (!invalidIndexes.isEmpty()) {
+                handleError(format(SHIPPING_ADDRESSES_ARE_NOT_VALID, customerDraft.getKey(), invalidIndexes));
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    @Nonnull
+    private List<Integer> getIndexValues(@Nonnull final List<Integer> list,
+        @Nonnull final Predicate<Integer> predicate) {
+
+        final List<Integer> indexes = new ArrayList<>();
+        for (Integer integer : list) {
+            if (predicate.test(integer)) {
+                indexes.add(integer);
+            }
+        }
+        return indexes;
+    }
+
+    private void collectReferencedKeys(
+        @Nonnull final CustomerBatchValidator.ReferencedKeys referencedKeys,
+        @Nonnull final CustomerDraft customerDraft) {
+
+        collectReferencedKeyFromResourceIdentifier(customerDraft.getCustomerGroup(),
+            referencedKeys.customerGroupKeys::add);
+        collectReferencedKeyFromCustomFieldsDraft(customerDraft.getCustom(),
+            referencedKeys.typeKeys::add);
+    }
+
+    public static class ReferencedKeys {
```

Comment:
It seems not necessary to make it `static`

Reply:
IMO in java inner classes are better with static vs non-static, otherwise every instance will create a different reference of it, also static inner class can not access it's enclosing scope, which is better in terms of secure, I guess there was a section about this in effective java book, maybe you might check and let me know if I am wrong.

Reply:
I get your point now. It's ok :)